components/tidb_query_common/src/storage/scanner.rs

```diff
@@ -23,6 +33,34 @@ pub struct RangesScanner<T> {
     current_range: IntervalRange,
     working_range_begin_key: Vec<u8>,
     working_range_end_key: Vec<u8>,
+    rescheduler: RescheduleChecker,
+}
+
+// TODO: maybe it's better to make it generic to avoid directly depending
+// on yatp's rescheduler.
+struct RescheduleChecker {
+    prev_start: Instant,
+    prev_key_count: usize,
+}
+
+impl RescheduleChecker {
+    fn new() -> Self {
+        Self {
+            prev_start: Instant::now(),
+            prev_key_count: 0,
+        }
+    }
+
+    async fn check_reschedule(&mut self, force_check: bool) {
```

Comment:
This function deserves careful compiler hints like `likely` and `inline(always)`, as it will be invoked for each row and ensuring its performance is critical.

Reply:
> > The only drawback is that big task will costs more time to finish
> 
> How bad is the performance drop for large queries?

I my benchmark, the P999 latency increase 3 times from 100ms to 300ms because after this change it needs about 50 times of reschedule to finish while before is less than 5 times.  But the benefit is that the throughput increased by 2 times from about 1000qps to 3000+qps.